src/app/clusters/chime-server/chime-server.cpp

```diff
@@ -0,0 +1,185 @@
+using namespace chip::app::Clusters::Chime::Attributes;
+using chip::Protocols::InteractionModel::Status;
+
+namespace chip {
+namespace app {
+namespace Clusters {
+
+ChimeServer::ChimeServer(EndpointId endpointId, ChimeDelegate & delegate) :
+    AttributeAccessInterface(MakeOptional(endpointId), Chime::Id), CommandHandlerInterface(MakeOptional(endpointId), Chime::Id),
+    mDelegate(delegate)
+{
+    mDelegate.SetChimeServer(this);
+}
+
+ChimeServer::~ChimeServer()
+{
+    AttributeAccessInterfaceRegistry::Instance().Unregister(this);
+    CommandHandlerInterfaceRegistry::Instance().UnregisterCommandHandler(this);
+}
+
+CHIP_ERROR ChimeServer::Init()
+{
+    VerifyOrReturnError(AttributeAccessInterfaceRegistry::Instance().Register(this), CHIP_ERROR_INTERNAL);
+    ReturnErrorOnFailure(CommandHandlerInterfaceRegistry::Instance().RegisterCommandHandler(this));
+    return CHIP_NO_ERROR;
+}
+
+// AttributeAccessInterface
+CHIP_ERROR ChimeServer::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    VerifyOrDie(aPath.mClusterId == Chime::Id);
+
+    switch (aPath.mAttributeId)
+    {
+    case InstalledChimeSounds::Id:
+        return aEncoder.EncodeList([this](const auto & encoder) -> CHIP_ERROR {
+            uint8_t index  = 0;
+            CHIP_ERROR err = CHIP_NO_ERROR;
+
+            // runs through all the Chime sounds available one by one
+            do
+            {
+                Chime::Structs::ChimeSoundStruct::Type chimeSound;
+                if ((err = this->mDelegate.GetChimeSoundByIndex(index, chimeSound)) == CHIP_NO_ERROR)
+                {
+                    ReturnErrorOnFailure(encoder.Encode(chimeSound));
+                    index++;
+                }
+            } while (err == CHIP_NO_ERROR);
+
+            if (err == CHIP_ERROR_NOT_FOUND)
+            {
+                return CHIP_NO_ERROR;
+            }
+            return err;
+        });
+
+    case ActiveChimeID::Id:
+        return aEncoder.Encode(mDelegate.GetActiveChimeId());
+
+    case Enabled::Id:
+        return aEncoder.Encode(mDelegate.GetEnabled());
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ChimeServer::Write(const ConcreteDataAttributePath & aPath, AttributeValueDecoder & aDecoder)
+{
+    VerifyOrDie(aPath.mClusterId == Chime::Id);
+
+    switch (aPath.mAttributeId)
+    {
+    case ActiveChimeID::Id: {
+        uint8_t newValue;
+        ReturnErrorOnFailure(aDecoder.Decode(newValue));
+        ReturnErrorOnFailure(mDelegate.SetActiveChimeId(newValue));
+        return CHIP_NO_ERROR;
+    }
+    case Enabled::Id: {
+        bool newValue;
+        ReturnErrorOnFailure(aDecoder.Decode(newValue));
+        ReturnErrorOnFailure(mDelegate.SetEnabled(newValue));
+        return CHIP_NO_ERROR;
+    }
+
+    default:
+        // Unknown attribute
+        return CHIP_IM_GLOBAL_STATUS(UnsupportedAttribute);
+    }
+}
+
+CHIP_ERROR ChimeServer::SetActiveChimeId(uint8_t soundId)
```

Comment:
So this is a private method... who calls it?

Reply:
Error, fixed.